cmd/panfs-multipart.go

```diff
@@ -73,68 +106,125 @@ func (fs *PANFSObjects) decodePartFile(name string) (partNumber int, etag string
 		return
 	}
-
+	fs.appendFileMapMu.Lock()
 	file := fs.appendFileMap[uploadID]
 	if file == nil {
-		file = &panfsAppendFile{
-			filePath: pathJoin(bucketPath, panfsS3TmpDir, fs.nodeDataSerial, bgAppendsDirName, fmt.Sprintf("%s.%s", uploadID, mustGetUUID())),
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
+		if file == nil {
+			fs.appendFileMapMu.Unlock()
+			return
 		}
 		fs.appendFileMap[uploadID] = file
 	}
 	fs.appendFileMapMu.Unlock()
 
-	file.Lock()
-	defer file.Unlock()
+	// No need to take a lock when bgAppend called from CompleteMultipartUpload which has been already taken the lock
+	if needFlock {
+		if !file.flock.TryLock() {
+			return
+		}
+		defer file.flock.Unlock()
+	}
 
-	// Since we append sequentially nextPartNumber will always be len(file.parts)+1
-	nextPartNumber := len(file.parts) + 1
-	uploadIDDir := fs.getUploadIDDir(bucketPath, bucket, object, uploadID)
+	uploadIDDir := fs.getUploadIDDir(bucketPath, object, uploadID)
 
-	entries, err := readDir(uploadIDDir)
+	// Read initial mod time of upload dir
+	prevUploadIDDirStat, err := fsStatDir(ctx, uploadIDDir)
 	if err != nil {
-		logger.GetReqInfo(ctx).AppendTags("uploadIDDir", uploadIDDir)
 		logger.LogIf(ctx, err)
 		return
 	}
-	sort.Strings(entries)
+	fsParts := fs.readMPartUploadParts(bucketPath, object, uploadID)
+	// Since we append sequentially nextPartNumber will always be len(fsParts.Appended)+1
+	initialAppendedLen := len(fsParts.Parts)
+	nextPartNumber := initialAppendedLen + 1
 
-	for _, entry := range entries {
-		if entry == fs.metaJSONFile {
-			continue
-		}
-		partNumber, etag, actualSize, err := fs.decodePartFile(entry)
+	for {
+		entries, err := readDir(uploadIDDir)
 		if err != nil {
-			// Skip part files whose name don't match expected format. These could be backend filesystem specific files.
-			continue
-		}
-		if partNumber < nextPartNumber {
-			// Part already appended.
-			continue
-		}
-		if partNumber > nextPartNumber {
-			// Required part number is not yet uploaded.
+			logger.GetReqInfo(ctx).AppendTags("uploadIDDir", uploadIDDir)
+			logger.LogIf(ctx, err)
 			return
 		}
+		sort.Strings(entries)
+
+		for _, entry := range entries {
+			if entry == fs.metaJSONFile {
+				continue
+			}
+			partNumber, etag, actualSize, err := fs.decodePartFile(entry)
+			if err != nil {
+				// Skip part files whose name don't match expected format. These could be backend filesystem specific files.
+				continue
+			}
+			if partNumber < nextPartNumber {
+				// Part already appended.
+				continue
+			}
+			if partNumber > nextPartNumber {
+				// Required part number is not yet uploaded.
+				break
+			}
+
+			partPath := pathJoin(uploadIDDir, entry)
+			err = xioutil.AppendFile(file.filePath, partPath, globalFSOSync)
+			if err != nil {
+				reqInfo := logger.GetReqInfo(ctx).AppendTags("partPath", partPath)
+				reqInfo.AppendTags("filepath", file.filePath)
+				logger.LogIf(ctx, err)
+				return
```

Comment:
I did not actually get your point on removing `mparts.json` - what the goal of the removal? bgAppend will update `mparts.json` at the end of appending in any case.
As for return from appending code - that is makes sense. Maybe in case of any error during append procedure we can remove append file and `mparts.json` to do not have corrupted appended file. The next put part or complete will append all parts from scratch

Reply:
The root cause of the problem reported by `xioutil.AppendFile()` on line 177 may be make it impossible to remove the `mparts.json` file during error handling. This way we might end up in a state where during the next call to `backgroundAppend()` the contents of the append file do not correspond to the state described by `mparts.json`.